src/main/native/CCM.c

```diff
@@ -267,12 +265,12 @@ JNIEXPORT jint JNICALL Java_com_ibm_crypto_plus_provider_ock_NativeInterface_do_
         handleIV_CCM(ivLen, keyLen, blockSize, J0Offset, (char*) iv, (char*) (parm_block + keyOffset), (char*) parm_block);
         memcpy(parm_block + counterValueOffset, parm_block + J0Offset + blockSize - 4, 4); // Add Counter Value
 
-        CCMzS390(input, output, aad, &len, &alen, parm_block, &mode);
+        CCMzS390((input != NULL) ? (input + inputOffset) : NULL, (output != NULL) ? (output + outputOffset) : NULL, aad, &len, &alen, parm_block, &mode);
 
-        if(inputJ != NULL) (*env)->ReleasePrimitiveArrayCritical(env, inputJ, input, 0);
-        if(outputJ != NULL) (*env)->ReleasePrimitiveArrayCritical(env, outputJ, output, 0);
+        if(inputJ != NULL) (*env)->ReleasePrimitiveArrayCritical(env, inputJ, (input != NULL) ? (input + inputOffset) : NULL, 0);
```

Comment:
Why does the `ReleasePrimitiveArrayCritical` have different parameters than the `GetPrimitiveArrayCritical`? Shouldn't we just release `input`?

Also, should we check `if(input != NULL)` instead of `if(inputJ != NULL)`?

Reply:
I mean that I think it should be `if(input != NULL) (*env)->ReleasePrimitiveArrayCritical(env, inputJ, input, 0);`, since the get and release should be applied to the same arguments, not one with offset.

Reply:
Okay, perhaps I'm missing something here then. The `GetPrimitiveArrayCritical` returns `input`, and does not take it as an argument. In the original code, by the time `ReleasePrimitiveArrayCritical` is called, `input` already has the offset applied. If we don't include the offset here, then that would be a functional change to the code.

Reply:
The offset is applied to read starting from a different part of the array rather than the start. But the full array is loaded and we get a pointer in `input`. So, when releasing that array, the full one should be released. At least to my understanding, the `inputOffset` is just for the reading part. The array, however, should be acquired and released as a whole.

I don't think it will affect anything if `input` is not a copy of the original array, but `GetPrimitiveArrayCritical` cannot always ensure that that's the case. So, if it is a copy we need to free the whole array and not just the part that starts at `inputOffset`.

@jasonkatonica Do you have any further insight on this? Maybe I'm wrong on the use of

Reply:
From what i can tell it does seem like there is another potential leak here in the case where `input` really created a copy when `GetPrimitiveArrayCritical` was called above. I thihnk we should be freeing all memory obtained by `GetPrimitiveArrayCritical` represented by the pointer input if the value is not null. There is other code spinkled through out the other C code that follows a pattern like this instead of trying to free at an offset of the input and output pointers.